core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournal.java

```diff
@@ -14,133 +14,384 @@
    *
    * @param location the location for this journal
    */
   public UfsJournal(URI location) {
-    mLocation = location;
-    mJournalFormatter = JournalFormatter.Factory.create();
+    this(location, UnderFileSystem.Factory.get(location.toString()));
   }
 
   /**
-   * @return the location of the completed logs
+   * Creates a new instance of {@link UfsJournal}.
+   *
+   * @param location the location for this journal
+   * @param ufs the under file system
    */
-  public URI getCompletedLocation() {
-    try {
-      return URIUtils.appendPath(mLocation, COMPLETED_LOCATION);
-    } catch (URISyntaxException e) {
-      throw new RuntimeException(e);
+  UfsJournal(URI location, UnderFileSystem ufs) {
+    mLocation = URIUtils.appendPathOrDie(location, VERSION);
+    mUfs = ufs;
+
+    mLogDir = URIUtils.appendPathOrDie(mLocation, LOG_DIRNAME);
+    mCheckpointDir = URIUtils.appendPathOrDie(mLocation, CHECKPOINT_DIRNAME);
+    mTmpDir = URIUtils.appendPathOrDie(mLocation, TMP_DIRNAME);
+  }
+
+  @Override
+  public URI getLocation() {
+    return mLocation;
+  }
+
+  @Override
+  public JournalReader getReader(JournalReaderOptions options) {
+    return new UfsJournalReader(this, options);
+  }
+
+  @Override
+  public JournalWriter getWriter(JournalWriterOptions options) throws IOException {
+    if (options.isPrimary()) {
+      return new UfsJournalLogWriter(this, options);
+    } else {
+      return new UfsJournalCheckpointWriter(this, options);
+    }
+  }
+
+  @Override
+  public boolean isFormatted() throws IOException {
+    UnderFileSystem ufs = UnderFileSystem.Factory.get(mLocation.toString());
+    UnderFileStatus[] files = ufs.listStatus(mLocation.toString());
+    if (files == null) {
+      return false;
     }
+    // Search for the format file.
+    String formatFilePrefix = Configuration.get(PropertyKey.MASTER_FORMAT_FILE_PREFIX);
+    for (UnderFileStatus file : files) {
+      if (file.getName().startsWith(formatFilePrefix)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public void format() throws IOException {
+    URI location = getLocation();
+    LOG.info("Formatting {}", location);
+    if (mUfs.isDirectory(location.toString())) {
+      for (UnderFileStatus p : mUfs.listStatus(location.toString())) {
+        URI childPath = URIUtils.appendPathOrDie(location, p.getName());
+        boolean failedToDelete;
+        if (p.isDirectory()) {
+          failedToDelete = !mUfs
+              .deleteDirectory(childPath.toString(), DeleteOptions.defaults().setRecursive(true));
+        } else {
+          failedToDelete = !mUfs.deleteFile(childPath.toString());
+        }
+        if (failedToDelete) {
+          throw new IOException(String.format("Failed to delete %s", childPath));
+        }
+      }
+    } else if (!mUfs.mkdirs(location.toString())) {
+      throw new IOException(String.format("Failed to create %s", location));
+    }
```

Comment:
@gpang  I am not sure. I copied the old format logic. I guess it was written this way because we might not have permission to delete the directory? 
I think it is common for admin to create a journal directory manually.